.github/workflows/yocto-build-deploy.yml

```diff
@@ -0,0 +1,676 @@
+          SLUG: "${{ steps.balena-lib.outputs.device_slug }}"
+          APPNAME: "${{ steps.balena-lib.outputs.device_slug }}"
+          BALENAOS_TOKEN: "${{ secrets.BALENA_API_KEY }}"
+          META_BALENA_VERSION: "${{ steps.balena-lib.outputs.meta_balena_version }}"
+          RELEASE_VERSION: "${{ steps.balena-lib.outputs.os_version }}"
+          BOOTABLE: 1
+          DEPLOY: yes
+          FINAL: yes
+          ESR: "${{ inputs.deploy-esr }}"
+          balenaCloudEmail:
+          balenaCloudPassword:
+          HOSTAPP_PATH: ${{ runner.temp }}/deploy/balena-image.docker
+        run: |
+          docker run --rm -t \
+            --privileged \
+            -e APPNAME \
+            -e API_ENV \
+            -e BALENAOS_TOKEN \
+            -e BALENAOS_ACCOUNT \
+            -e META_BALENA_VERSION \
+            -e RELEASE_VERSION \
+            -e MACHINE \
+            -e VERBOSE \
+            -e BOOTABLE \
+            -e DEPLOY \
+            -e FINAL \
+            -e ESR \
+            -e SECURE_BOOT_FEATURE_FLAG \
+            -e balenaCloudEmail \
+            -e balenaCloudPassword \
+            -v "$(readlink --canonicalize "${HOSTAPP_PATH}")":/host/appimage.docker \
+            -v "${WORKSPACE}":/work \
+            -v "${WORKSPACE}":/deploy \
+            "${{ steps.hostapp-helper-image.outputs.id }}" /balena-deploy-block.sh
+
+      ##############################
+      # AMI Deploy
+      ##############################
+
+      - name: Set AMI arch
+        id: ami-arch
+        if: inputs.deploy-ami == true
+        run: |
+          if [ "${dt_arch}" = "amd64" ]; then
+            echo "string=x86_64" >> $GITHUB_OUTPUT
+          elif [ "${dt_arch}" = "aarch64" ]; then
+            echo "string=arm64" >> $GITHUB_OUTPUT
+          fi
+
+      # AMI name format: balenaOS(-installer?)(-secureboot?)-VERSION-DEVICE_TYPE
+      - name: Set AMI name
+        id: ami-name
+        if: inputs.deploy-ami == true
+        run: |
+          if [ "${{ inputs.sign-image }}" = "true" ]; then
+            echo "string=balenaOS-secureboot-${VERSION}-${MACHINE}" >> $GITHUB_OUTPUT
+          else
+            echo "string=balenaOS-${VERSION}-${MACHINE}" >> $GITHUB_OUTPUT
+          fi
+
+      - name: Pull helper image
+        id: ami-helper-image
+        if: inputs.deploy-ami == true
+        env:
+          HELPER_IMAGE_REPO: ghcr.io/balena-os/balena-yocto-scripts
+          YOCTO_SCRIPTS_VERSION: ${{ steps.balena-lib.outputs.yocto_scripts_version }}
+          YOCTO_SCRIPTS_REF: ${{ steps.balena-lib.outputs.yocto_scripts_ref }}
+          HELPER_IMAGE_VARIANT: yocto-build-env
+        run: |
+          image_tag="${HELPER_IMAGE_REPO}:${YOCTO_SCRIPTS_VERSION}-${HELPER_IMAGE_VARIANT}"
+          if ! docker pull "${image_tag}"; then
+            image_tag="${HELPER_IMAGE_REPO}:${YOCTO_SCRIPTS_REF}-${HELPER_IMAGE_VARIANT}"
+            docker pull "${image_tag}"
+          fi
```

Comment:
Try to pull the helper image via the yocto-scripts repo semver.

For open PRs, this semver will not exist on final images yet, only draft images will exist, so fallback to the yocto-scripts git ref. This way we are testing changes to Dockerfiles as part of this PR.

The downside being that the Dockerfiles are untrusted user code and can be modified by forks, which is insecure and relies on human judgement to notice.

Reply:
Actually moving the logic into the workflow file instead, its not that bad once simplified (as we know the state we're going to be in once we get to this point)